### DISCUSSION TITLE
Dynamic JS function imports stopped working when migration to Next 15

### DISCUSSION THREAD

**Discussion by PILLOWPET** [vercel/next.js#71903] [Help]
### Summary

I am attempting to migrate our stack from Next 14.1 to Next 15.0.1
We have an architecture where api client fetchers are both used by SWR and backend fetchers:
- `useFoo.ts`
```
import { fetchFoo } from './foo-fetcher`
export const useFoo = (
): SWRResponse =>
  useSWR(
    '/foo'
    () => fetchFoo(false)
  );
```

- `fetchFoo.ts`

```
import { fetchApi } from './baseApiClient';

export const fetchFoo = async (
  serverSide = false,
) =>
  await fetchApi('/foo' , {
    throwOnError: true,
    serverSide,
  });
```

- `baseApiClient.ts`

```
export const fetchApi = (
  apiSchema: any,
  options?: any,
) => {
  const {
    serverSide = false,
    throwOnError = false,
  } = options ?? {};

  // Line causing the issue
  const authHeaders = serverSide
    ? // Can only be used server side
      await (await import('./serverApiHelpers')).getServerAuthHeaders()
    : undefined;

  const path = getApiPath(apiSchema.path);

  const response = await fetch(path, {
    method: apiSchema.method,
    headers: {
      ...authHeaders,
    },
    body:{},
  });

  return response;
};
```

The `getServerAuthHeaders` uses a lib that has an `server-only` line at the top of the file, meaning it should only be used server side. I don't understand why since `serverSide` is `false` when I use SWR, it still attempts to import `serverApiHelpers` when calling the `useFoo` hook, even with the dynamic import on ` (await import('./serverApiHelpers')).getServerAuthHeaders()` 

I get the following error:

```
'server-only' cannot be imported from a Client Component module. It should only be used from a Server Component.

The error was caused by importing '@clerk/nextjs/dist/esm/server/index.js' in './src/backend/utils/auth.ts'.

Import trace for requested module:
./src/backend/utils/auth.ts

./src/adapters/api-client/serverApiHelpers.ts

./src/adapters/api-client/baseApiClient.ts

./src/adapters/api-client/fetchers/fetchFoo.ts

./src/adapters/api-client/client-fetchers/useFoo.ts

./src/components/Foo.tsx
```

### Additional information

```js
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 24.0.0: Tue Sep 24 23:37:25 PDT 2024; root:xnu-11215.1.12~1/RELEASE_ARM64_T6030
  Available memory (MB): 18432
  Available CPU cores: 12
Binaries:
  Node: 20.12.2
  npm: 10.5.0
  Yarn: 1.22.22
  pnpm: 8.15.5
Relevant Packages:
  next: 15.0.1
  eslint-config-next: 15.0.1
  react: 19.0.0-rc-69d4b800-20241021
  react-dom: 19.0.0-rc-69d4b800-20241021
  typescript: 5.0.4
```
```


### Example

_No response_

**Comment by Abderrahmanlahmidi** [comment]
Instead of having the authHeaders initialization inside fetchApi, try isolating the server-side functionality outside the main function and only call it when serverSide is true.

Create a helper for conditional importing.

typescript
Copy code
// create getServerAuthHeaders.ts or a similar helper
export const getServerAuthHeaders = async () => {
  if (typeof window === 'undefined') {
    const { getServerAuthHeaders } = await import('./serverApiHelpers');
    return getServerAuthHeaders();
  }
  return undefined;
};
Update fetchApi to use this helper.

typescript
Copy code
export const fetchApi = async (
  apiSchema: any,
  options?: any,
) => {
  const { serverSide = false, throwOnError = false } = options ?? {};

  // Use helper for conditional import
  const authHeaders = serverSide ? await getServerAuthHeaders() : undefined;
  const path = getApiPath(apiSchema.path);

  const response = await fetch(path, {
    method: apiSchema.method,
    headers: { ...authHeaders },
    body: {},
  });

  return response;
};
Solution 2: Separate Client and Server Fetch Functions
Split fetchApi into two distinct functions: one for server-side usage and one for client-side usage.

Create fetchApiServer and fetchApiClient.

Server-side function includes getServerAuthHeaders.
Client-side function omits the server-specific import entirely.
Update useFoo to call the appropriate function based on the serverSide parameter.

Solution 3: Use Middleware or API Routes for Server-Side Logic
If you need to keep your code universal, consider shifting server-specific logic to an API route or middleware within Next.js. This way, your client component can make a request to an internal API endpoint that handles getServerAuthHeaders logic securely on the server side.

Summary
These approaches should help isolate the server-only imports and prevent the bundler from trying to import them in client-side contexts, resolving the server-only import error. Solution 1 is the most straightforward, but consider Solution 2 or 3 if you foresee more client-server splits in your fetch logic. Let me know if you need more details on any of these steps!

**Comment by PILLOWPET** [comment]
Thanks for your help,

For solution 1, it is actually what is implemented in the code so far. I simplified it for a better understanding of the code but the code actually looks like this in the `baseApiClient.ts`

```

const getAuthHeaders = async (serverSide: boolean) =>
  serverSide
    ? // Can only be used server side
      await (await import('./serverApiHelpers')).getServerAuthHeaders()
    : undefined;


export const fetchApi = (
  apiSchema: any,
  options?: any,
) => {
  const {
    serverSide = false,
    throwOnError = false,
  } = options ?? {};

  // Line causing the issue
  const authHeaders = await getAuthHeaders(serverSide);

  const path = getApiPath(apiSchema.path);

  const response = await fetch(path, {
    method: apiSchema.method,
    headers: {
      ...authHeaders,
    },
    body:{},
  });

  return response;
};
```

I tried adding the check on window in the `getAuthHeaders` function but it has not changed anything.

We might split the fetchers in our API routes but this is a hassle considering the amount of swr/fetchers we have. Furthermore, I'm not sure I understand why the dynamic import is always executed regardless of the value of `serverSide`.

**Comment by aniruddhaadak80** [comment]
Hi @PILLOWPET ,

It looks like you're experiencing an issue with dynamic imports after migrating to Next.js 15. The error indicates that a server-only import is being attempted from a client component. Here are a few suggestions to help resolve the issue:

1. **Ensure Correct Imports**: Verify that `getServerAuthHeaders` and any other server-only code are not being used directly in client components. Make sure that they are only called within server-side code or components.

2. **Conditional Import**: If you need to conditionally import server-only functions, ensure that this logic is placed in a server-side context. For instance, consider moving the import logic inside a server-side function or a server-only API route.

3. **Check Usage in SWR**: Since `useFoo` is called in a client component, make sure that `fetchApi` does not attempt to import server-only modules when used in that context. You might want to split the logic between client and server API calls.

4. **Dynamic Import**: If you want to keep the dynamic import, ensure it only executes in a server context. For example:
   ```javascript
   let authHeaders;
   if (serverSide) {
     authHeaders = await import('./serverApiHelpers').then(module => module.getServerAuthHeaders());
   }
   ```

By following these steps, you should be able to prevent the server-only import error while still maintaining your application's functionality. If you have further questions or need additional assistance, let me know. Thanks!
